src/OpenIddict.Mvc/OpenIddictController.cs

```diff
@@ -15,23 +15,20 @@
 using Microsoft.AspNetCore.Authorization;
 using Microsoft.AspNetCore.Builder;
 using Microsoft.AspNetCore.Http.Authentication;
+using Microsoft.AspNetCore.Identity;
 using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Options;
 using Microsoft.IdentityModel.Protocols.OpenIdConnect;
 
 namespace OpenIddict.Mvc {
     // Note: this controller is generic and doesn't need to be marked as internal to prevent MVC from discovering it.
-    public class OpenIddictController<TUser, TApplication> : Controller where TUser : class where TApplication : class {
-        public OpenIddictController([NotNull] OpenIddictServices<TUser, TApplication> services) {
-            Services = services;
-        }
-
-        /// <summary>
-        /// Gets the OpenIddict services used by the controller.
-        /// </summary>
-        protected virtual OpenIddictServices<TUser, TApplication> Services { get; }
-
+    public class OpenIddictController<TUser, TApplication, TAuthorization, TToken> : Controller
```

Comment:
Because you don't like the service locator pattern? :smile: 
I tried to get rid of it everywhere I could. `OpenIddictServices` is now only used in the provider itself (that's why it's in the `Infrastructure` namespace)